practical-k-means/practical-k-means-cancer-gene-expression.ipynb

```diff
@@ -0,0 +1,482 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# K-Means Clustering in Python -- TCGA Cancer Gene Expression"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "This jupyter notebook accompanies the Real Python's [Practical K-Means Clustering in Python](https://realpython.com/practical-k-means-clustering-python/) article. Specifically, this notebook works through the **second half** of the article, from where the TCGA cancer gene expression dataset is introduced."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Building a K-Means Clustering Pipeline"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import tarfile\n",
+    "import urllib\n",
+    "\n",
+    "import numpy as np\n",
+    "import matplotlib.pyplot as plt\n",
+    "import pandas as pd\n",
+    "import seaborn as sns\n",
+    "\n",
+    "from sklearn.cluster import KMeans\n",
+    "from sklearn.decomposition import PCA\n",
+    "from sklearn.metrics import silhouette_score, adjusted_rand_score\n",
+    "from sklearn.pipeline import Pipeline\n",
+    "from sklearn.preprocessing import LabelEncoder, MinMaxScaler"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "uci_tcga_url = \"https://archive.ics.uci.edu/ml/machine-learning-databases/00401/\"\n",
+    "archive_name = \"TCGA-PANCAN-HiSeq-801x20531.tar.gz\"\n",
+    "\n",
+    "# build the url\n",
+    "full_download_url = urllib.parse.urljoin(uci_tcga_url, archive_name)\n",
+    "\n",
+    "# download the file\n",
+    "r = urllib.request.urlretrieve (full_download_url, archive_name)\n",
+    "\n",
+    "# extract the data from the archive\n",
+    "tar = tarfile.open(archive_name, \"r:gz\")\n",
+    "tar.extractall()\n",
+    "tar.close()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "datafile = \"TCGA-PANCAN-HiSeq-801x20531/data.csv\"\n",
+    "labels_file = \"TCGA-PANCAN-HiSeq-801x20531/labels.csv\"\n",
+    "\n",
+    "data = np.genfromtxt(\n",
+    "    datafile,\n",
+    "    delimiter=\",\",\n",
+    "    usecols=range(1, 20532),\n",
+    "    skip_header=1\n",
+    ")\n",
+    "\n",
+    "true_label_names = np.genfromtxt(\n",
+    "    labels_file,\n",
+    "    delimiter=\",\",\n",
+    "    usecols=(1,),\n",
+    "    skip_header=1,\n",
+    "    dtype="str"\n",
```

Comment:
@arvkevi I found the issue with this notebook. If I fix this line, I can render the notebook. I already checked and the code looks good in it:

```suggestion
    "    dtype=str\n",
```

Reply:
Wow, good find, I don't know how long that would have taken me to figure out!

Reply:
Haha, thanks @arvkevi! I knew from the error that I got that it had to be malformed JSON, so I just took the raw notebook JSON and copied/pasted it into https://jsonlint.com/ to see if it could flag any trouble spots. Lo and behold it found that spot for me!